src/delphi/train/dataset_tokenization.py

```diff
@@ -0,0 +1,50 @@
+from collections import deque
+
+from transformers import PreTrainedTokenizerBase
+
+
+def extend_deque(
+    dq: deque[int],
+    context_size: int,
+    text_stories: list[str],
+    prompt_idx: int,
+    tokenizer: PreTrainedTokenizerBase,
+) -> int:
+    while len(dq) < context_size and prompt_idx < len(text_stories):
+        text_story = text_stories[prompt_idx]
+        dq.extend(
+            tokenizer.encode(text_story, add_special_tokens=False)
+            + [tokenizer.eos_token_id]
+        )
+        prompt_idx += 1
+    return prompt_idx
+
+
+def make_new_samples(
+    dq: deque[int], context_size: int, tokenizer: PreTrainedTokenizerBase
+) -> list[list[int]]:
+    samples = []
+    while len(dq) >= context_size:
+        sample = [tokenizer.bos_token_id]
+        for _ in range(context_size - 1):  # peek at and not pop the last element
+            sample.append(dq.popleft())
+        sample.append(dq[0])
+        samples.append(sample)
+    return samples
+
+
+def get_tokenized_batches(
```

Comment:
@siwei-li , a reply to my questions would have been much appreciated, instead of just resolving them. They might be simple or even stupid questions, but it would have helped me understand your code.

Reply:
there are no batches, so I think the function name should be changed? to something like "tokenize_dataset"